### ISSUE TITLE
Precompile fails on Julia 0.6

### ISSUE THREAD

**Issue by pintergreg** [c-s/DataCubes.jl#3]
I could not have used DataCubes with Julia 0.6, because the precompiling fails. Output is below:
```
julia> using DataCubes
INFO: Precompiling module DataCubes.

WARNING: deprecated syntax "inner constructor LDict(...) around /home/gergo/.julia/v0.6/DataCubes/src/datatypes/ldict.jl:22".
Use "LDict{K,V}(...) where {K,V}" instead.

WARNING: deprecated syntax "inner constructor LDict(...) around /home/gergo/.julia/v0.6/DataCubes/src/datatypes/ldict.jl:23".
Use "LDict{K,V}(...) where {K,V}" instead.

WARNING: deprecated syntax "abstract AbstractLabeledArray{T,N,AXES,TN}<:AbstractArray{T,N}" at /home/gergo/.julia/v0.6/DataCubes/src/datatypes/abstract_types.jl:2.
Use "abstract type AbstractLabeledArray{T,N,AXES,TN}<:AbstractArray{T,N} end" instead.

WARNING: deprecated syntax "abstract AbstractDictArray{K,N,VS,SV}<:AbstractArray{LDict{K,SV},N}" at /home/gergo/.julia/v0.6/DataCubes/src/datatypes/abstract_types.jl:3.
Use "abstract type AbstractDictArray{K,N,VS,SV}<:AbstractArray{LDict{K,SV},N} end" instead.

WARNING: deprecated syntax "inner constructor DictArray(...) around /home/gergo/.julia/v0.6/DataCubes/src/datatypes/dict_array.jl:36".
Use "DictArray{K,N,VS,SV}(...) where {K,N,VS,SV}" instead.
ERROR: LoadError: LoadError: ArgumentError: method definition for getindex at /home/gergo/.julia/v0.6/DataCubes/src/datatypes/dict_array.jl:167 has free type variables
Stacktrace:
 [1] anonymous at ./<missing>:2
while loading /home/gergo/.julia/v0.6/DataCubes/src/datatypes/dict_array.jl, in expression starting on line 167
while loading /home/gergo/.julia/v0.6/DataCubes/src/DataCubes.jl, in expression starting on line 13
ERROR: Failed to precompile DataCubes to /home/gergo/.julia/lib/v0.6/DataCubes.ji.
Stacktrace:
 [1] macro expansion at ./REPL.jl:97 [inlined]
 [2] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
```

**Comment by c-s** [comment]
Hi pintergreg,

I am working on upgrading DataCubes to be compatible with Julia v0.6. It seems it is much harder than the time when I upgraded DataCubes for Julia v0.5 from 0.4. I spent the last week fixing many issues related to the update, but there are still some to deal with. At this point, I try to pass all unit tests. I will keep you updated. Thanks!

**Comment by pintergreg** [comment]
@c-s I've just read about FemtoCleaner that is supposed to help with the compatibility issues. I works as a github app.
https://juliacomputing.com/blog/2017/08/17/femtocleaner.html
Note that I have zero experience with this tool, I've just found this post on reddit. :)

**Comment by c-s** [comment]
@pintergreg Thanks for the info! I just updated DataCubes to version 0.2.0, which is compatible with Julia v0.6. There are too many incompatibilities between Julia v0.5 and v0.6, mainly because DataCubes exploited some of the obscure rules for types and macros in Julia, and that's why it took so many days to upgrade. Anyway, it runs fine mostly. The only issue I experience is for some reason a .* b not converted into broadcast(*, a, b) properly (and so is for other arithmetic operators), so I had to use broadcast instead of . in many unit test cases. Thanks.
